### ISSUE TITLE
Setting a union type is completely busted

### ISSUE THREAD

**Issue by Shinmera** [cffi/cffi#383]
Reproducer:
```lisp
(cffi:defcunion foo
  (a :int32)
  (b :string))

(cffi:with-foreign-object (a '(:union foo))
  (setf (cffi:mem-ref a '(:union foo)) #xDEAD)
  (cffi:mem-ref a :int32))
; => garbage
```

It seems like setf-mem-ref is expanded with `expand-into-memory`, which does not have a specialised method for union types, and the default method just returns the value (which seems completely wrong to me), so this turns into a no-op, having `a` contain stack garbage.

Gotta say, finding this bug was a nightmare :)

**Comment by phoe** [comment]
Is it valid to `(setf (cffi:mem-ref a '(:union foo)) #xDEAD)` if it's a type mismatch (pointer versus integer)?

I'd have expected assignment to one of union members (assigning something of type either `:int32` or `:string`), or do C-style union assignment that would copy the whole union, but this does look like a malformed call at a first glance. A C compiler would complain about incompatible types when assigning to type `union foo` from type `int` or something.

The default method silently no-oping smells like a big bug though.

**Comment by phoe** [comment]
Yes, and `(cffi:mem-ref a '(:union foo))` is of type `(:union foo)`. Different foreign types.

According to CFFI docs:

> A union is a structure in which all slots have an offset of zero. It is isomorphic to the C union. Therefore, you should use the usual foreign structure operations for accessing a union’s slots. 

Do you expect CFFI to recognize that your union has a `:int32` member and automatically translate the call into writing a `:int32` into memory? Such magic doesn't happen for structs, why should it happen for unions?

**Comment by Shinmera** [comment]
Because conceptually why wouldn't it

**Comment by phoe** [comment]
`#xDEAD` is a constant, which makes your case easy. In the general case of `(setf (cffi:mem-ref a '(:union foo)) x)`, you'd need to know the Lisp type and/or foreign type of `x` at compilation time to be able to expand the macro at all, which means a dependency on `trivial-cltl2:variable-information` or something equivalent.

And if you know the simple foreign type at compilation time, you can as well stuff it into `mem-ref` yourself.

Also, there's trippy cases like `(cffi:defcunion foo (a :int32) (b :int64))` that might get weird with endianness.

**Comment by luismbo** [comment]
In any case, CFFI shouldn't silently do nothing, which is what happened there IIUC.

**Comment by Shinmera** [comment]
> `#xDEAD` is a constant, which makes your case easy. In the general case of `(setf (cffi:mem-ref a '(:union foo)) x)`, you'd need to know the Lisp type and/or foreign type of `x` at compilation time to be able to expand the macro at all, which means a dependency on `trivial-cltl2:variable-information` or something equivalent.

Huh? no you don't. Just expand to an etypecase.

> And if you know the simple foreign type at compilation time, you can as well stuff it into `mem-ref` yourself.

It'd be cool if cffi did the etypecase for me.

> Also, there's trippy cases like `(cffi:defcunion foo (a :int32) (b :int64))` that might get weird with endianness.

Sure. Error in that case, then. But there's plenty of cases where it *can* do an expansion that *does* do what you'd expect it to.

**Comment by phoe** [comment]
OK, I see what you mean now. (Didn't expect a request for runtime dispatch from you, I was thinking of compile-time solutions only!)

What would be the naïve thing to do here? You could iterate through all slots of cstruct definition and construct a typecase where the first matching type gets to execute the actual memset for that type. (Does CFFI already have some sort of mapping facility between foreign types and Lisp types, or will that need to get written?) For a more sturdy solution, maybe check if the slot types are disjoint and signal a compile-time error otherwise.
